src/Workspaces/Core/Portable/Log/HistogramLogAggregator.cs

```diff
@@ -1,53 +1,95 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.Internal.Log
 {
     /// <summary>
     /// Defines a log aggregator to create a histogram
     /// </summary>
-    internal class HistogramLogAggregator : AbstractLogAggregator<LogAggregator.Counter>
+    internal sealed class HistogramLogAggregator : AbstractLogAggregator<HistogramLogAggregator.HistogramCounter>
     {
         private readonly int _bucketSize;
         private readonly int _maxBucketValue;
+        private readonly int _bucketCount;
 
         public HistogramLogAggregator(int bucketSize, int maxBucketValue)
         {
+            if (bucketSize <= 0 || maxBucketValue <= 0 || maxBucketValue % bucketSize != 0)
+            {
+                throw new ArgumentException();
+            }
+
             _bucketSize = bucketSize;
             _maxBucketValue = maxBucketValue;
+            _bucketCount = maxBucketValue / bucketSize + 1;
         }
 
-        public void IncreaseCount(decimal value)
+        protected override HistogramCounter CreateCounter()
         {
-            var counter = GetCounter(GetBucket(value));
-            counter.IncreaseCount();
+            return new HistogramCounter(_bucketSize, _maxBucketValue, _bucketCount);
         }
 
-        public int GetCount(decimal value)
+        public void IncreaseCount(object key, decimal value)
         {
-            if (TryGetCounter(GetBucket(value), out var counter))
-            {
-                return counter.GetCount();
-            }
-
-            return 0;
+            var counter = GetCounter(key);
+            counter.IncreaseCount(value);
         }
 
-        protected override LogAggregator.Counter CreateCounter()
+        internal sealed class HistogramCounter
         {
-            return new LogAggregator.Counter();
-        }
+            private readonly int[] _buckets;
```

Comment:
I don't know how much this matters, but this will allocate more than previous right?  Since it's now storing the entire array instead of just the bucket interval and max value.

It's going to be a really small amount though so I guess I don't care too much, unless we start using this logger instances in a lot of places.

Reply:
Yup, this would be wasteful if the data is sparse and/or there's a lot of buckets. For now I think it's fine, we can switch to a dictionary if needed in the future.